### ISSUE TITLE
How to convert a channel first Keras model to channel last model?

### ISSUE THREAD

**Issue by mrtpk123** [tensorflow/tensorflow#49186]
Hello,

I have a pre-trained Keras model (in h5 format) where all the layers operate on channel first data format. I want to convert this model to operate on the channel last data format (the default data format). 
Does TensorFlow support this out of the box? Any help or pointers on how to do this is much appreciated.

For some clarity, the current model summary looks like this:

    Model: "model"
    __________________________________________________________________________________________________
    Layer (type)                    Output Shape         Param #     Connected to
    ==================================================================================================
    input0 (InputLayer)             [(None, 3, 240, 320) 0
    __________________________________________________________________________________________________
    259_pad (ZeroPadding2D)         (None, 3, 242, 322)  0           input0[0][0]
    __________________________________________________________________________________________________
    259 (Conv2D)                    (None, 16, 120, 160) 432         259_pad[0][0]
    __________________________________________________________________________________________________
    260 (BatchNormalization)        (None, 16, 120, 160) 64          259[0][0]
    __________________________________________________________________________________________________
    261 (Activation)                (None, 16, 120, 160) 0           260[0][0]
    __________________________________________________________________________________________________
    262_pad (ZeroPadding2D)         (None, 16, 122, 162) 0           261[0][0]
    __________________________________________________________________________________________________
    262 (DepthwiseConv2D)           (None, 16, 120, 160) 144         262_pad[0][0]

As you can see, it's in the channel first format. I want to convert each layer in the model to operate on channel last format. So the ideal model summary will be as follows:

    Model: "model"
    __________________________________________________________________________________________________
    Layer (type)                    Output Shape         Param #     Connected to
    ==================================================================================================
    input0 (InputLayer)             [(None, 240, 320, 3) 0
    __________________________________________________________________________________________________
    259_pad (ZeroPadding2D)         (None, 242, 322, 3)  0           input0[0][0]
    __________________________________________________________________________________________________
    259 (Conv2D)                    (None, 120, 160, 16) 432         259_pad[0][0]
    __________________________________________________________________________________________________
    260 (BatchNormalization)        (None, 120, 160, 16) 64          259[0][0]
    __________________________________________________________________________________________________
    261 (Activation)                (None, 120, 160, 16) 0           260[0][0]
    __________________________________________________________________________________________________
    262_pad (ZeroPadding2D)         (None, 122, 162, 16) 0           261[0][0]
    __________________________________________________________________________________________________
    262 (DepthwiseConv2D)           (None, 120, 160, 16) 144         262_pad[0][0]

Thanks.

**Comment by ymodak** [comment]
You may try https://www.tensorflow.org/api_docs/python/tf/keras/backend/set_image_data_format
```python
tf.keras.backend.set_image_data_format('channels_last')
```

**Comment by mrtpk123** [comment]
Thank you @ymodak  for your resposne.

Yes, it didn't work. The model is pretrained and is in channel first format. Setting `tf.keras.backend.set_image_data_format('channels_first')`  makes the default ordering to be channels_first and didn't affect the loaded model.

**Comment by bhack** [comment]
Have you tried to resave your model as in https://github.com/tensorflow/models/issues/3167#issuecomment-371410527 ?

**Comment by mrtpk123** [comment]
Thank you for the pointer. 

I have only the trained model. From the comment (2nd part) @bhack  mentioned do I have to do "network surgery"?

> 2 .The above solution works when the training files were available. In the case where I had only the output of tf.estimator.export_savedmodel, I had to do a network surgery. I made a clone of the graph, in channel_last format, loaded it in and assigned all variables, values from the trained model. This new graph runs fine on CPU.

If yes, can you share a sample code on how it is done on a Conv2D layer?

Any help is really appreciated. Thank you.

**Comment by bhack** [comment]
I think that without the training model it could be a nice tutorial for Keras-io. Probably you could open a request at https://github.com/keras-team/keras-io

**Comment by bhack** [comment]
If you will track the request there please close this and leave a reference to the new ticket.

**Comment by mrtpk123** [comment]
I will open a request at keras-io, but I'm would be really grateful if you could add in an example on how to do this for just one layer (say, Conv2D).

**Comment by bhack** [comment]
I don't know if you are doing this for performance or not but e.g. for you Conv2d request if you see the internals there are some constrains on the format in different places or implicit/internal conversions. 
E.g. Search `NHWC` ad `NCHW` in this source code:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_ops.cc.

**Comment by mrtpk123** [comment]
Thank you.  This is new info.

Does this [line](https://github.com/tensorflow/tensorflow/blob/9d461da4cb0af2f737bbfc68cca3f6445f1ceb60/tensorflow/core/kernels/conv_ops.cc#L86) - `CHECK(data_format == FORMAT_NHWC) << "Generic conv implementation only "` at `conv_ops.cc` mean that internally Conv2D operates in NHWC? 

So, if I parse Conv2D layers in the model and set their channel_format as NHWC, will it give the same output? I was planning to transpose the weights as well while setting the channel format to NHWC.

**Comment by bhack** [comment]
That was for  generic and e.g. also on CPU you can see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_ops.cc#L246-L247

**Comment by mrtpk123** [comment]
I have raised an [issue](https://github.com/keras-team/keras-io/issues/467) at keras-io. Will wait for their response to convert the model.
Thank you for your help and pointers. Really helpful.
